orte/mca/state/orted/state_orted.c

```diff
@@ -90,6 +93,107 @@ static orte_state_cbfunc_t proc_callbacks[] = {
     track_procs
 };
 
+#if SPC_ENABLE == 1
+#include <sys/time.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+
+#define SPC_BUFFER_SIZE 1024
+
+static opal_event_t ompi_spc_event;
+
+void copy_file(int src_fd, char *dest_name)
+{
+    int num_chars, dest_fd;
+    char buffer[SPC_BUFFER_SIZE];
+
+    opal_output(0, "\nWriting data to %s\n", dest_name);
+
+    if(-1 == (dest_fd = open(dest_name, O_WRONLY | O_CREAT, S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH))) {
+        opal_output(0, "ERROR: Could not open destination file.\n");
+        goto copy_cleanup;
+    }
+
+    while((num_chars = read(src_fd, buffer, SPC_BUFFER_SIZE)) > 0) {
+        if(num_chars != write(dest_fd, buffer, num_chars)) {
+            opal_output(0, "ERROR: Write error.\n");
+            goto copy_cleanup;
+        }
+        if(num_chars == -1) {
+            opal_output(0, "ERROR: Read error.\n");
+            goto copy_cleanup;
+        }
+    }
+
+ copy_cleanup:
+    if(-1 == close(dest_fd)) {
+        opal_output(0, "ERROR: Failed to close output file.\n");
+    }
+    if(-1 == lseek(src_fd, 0, SEEK_SET)) {
+        opal_output(0, "ERROR: Failed to reset file pointer.  %s\n", strerror(errno));
+    }
+}
+
+int spc_event_cb()
+{
+    int rc, i, j;
+    char sm_file[128], filename[128];
+    orte_proc_t *pptr;
+
+    static char *shm_dir;
+    static int first_time = 1;
+    static int *fd;
+
+    orte_node_t *node;
+    node = (orte_node_t*)opal_pointer_array_get_item(orte_node_pool, ORTE_PROC_MY_NAME->vpid);
+
+    if(first_time) {
+        first_time = 0;
+
+        if(0 == access("/dev/shm", W_OK)) {
+            shm_dir = "/dev/shm";
+        } else {
+            shm_dir = opal_process_info.job_session_dir;
+        }
+
+        fd = (int*)malloc(node->num_procs * sizeof(int));
+
+        for(i = 0; i < node->num_procs; i++) {
+            fd[i] = -1;
+        }
+    }
+
+    for(i = 0; i < node->num_procs; i++) {
+        if(NULL != (pptr = (orte_proc_t*)opal_pointer_array_get_item(node->procs, i))) {
+            if(fd[i] == -1) {
+                rc = sprintf(sm_file, "%s" OPAL_PATH_SEP "spc_data.%s.%d.%d", shm_dir,
```

Comment:
I am no sure I understand what this striking similar code is doing in the 2 cases.